### PR TITLE
DimensionsPanel: Fix unexpected value decoding/encoding

### LINKED ISSUE #51619: Dimension Controls: Sliders jump to custom input and are therefore basically not usable anymore
### Description

The behavior of sliders seem to have changed in 16.0.
It's not possible to use them anymore, since they automatically jump to a custom input after selecting.

### Step-by-step reproduction instructions

1. Create a theme with custom spacing sizes
2. Try to use the slider

### Screenshots, screen recording, code snippet

https://github.com/WordPress/gutenberg/assets/17861443/fa3690ce-7100-402f-8c49-53a545372f7a

Here is my theme.json:
image
image

### Environment info

- WP 6.2.2
- GB 16
- Custom theme

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### LINKED ISSUE #51891: Custom spacing sizes do not work properly after latest update
### Description

I have custom spacing sizes in my theme.json:
```
{
    "settings": {
        "spacing": {
            "spacingSizes": [
                {
                    "name": "6",
                    "slug": "60",
                    "size": "2rem"
                },
                ...
            ]
        }
    }
}
```
When I add these custom spacing sizes to a group in the block editor, the frontend displays them as:
```
<div class="wp-block-group alignwide bricksy-travel-two-column-text has-global-padding is-layout-constrained wp-block-group-is-layout-constrained" style="padding-top:2rem;padding-right:var(--wp--preset--spacing--60);padding-bottom:2rem;padding-left:var(--wp--preset--spacing--60)">
```
For the top and bottom, it converts them to their actual units, while it keeps the variables from the right and left properties. The order here is dependent on which size I set first in the editor. Sometimes I get the actual units for the left and right. This is a new behavior that I noticed after updating the plugin.

### Step-by-step reproduction instructions

1. Update to latest version of Gutenberg.
2. Add custom spacing sizes to theme.json.
3. Add group block in block editor and add padding using custom spacing sizes.
4. See the display on the frontend.

### Screenshots, screen recording, code snippet

Screen Shot 2023-06-25 at 11 54 17 PM
Screen Shot 2023-06-25 at 11 54 04 PM

### Environment info

- WordPress 6.2.2
- Gutenberg 16.0.0

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by t-hamano**
Try to: Fixes: #51619

- Fixes: #51619
- Fixes: #51891

## What?

This PR attempts to fix #51619 and #51891, regressions caused by #50956.

## Why?

#50956 took the following search and approach to fix the issue reported in #50940 where the `SpacingSizesControl` value was reset when the global style was saved.

- The value that `DiemsionsPanel` is trying to decode (padding/margin) is an object with `top/bottom/left/right` properties
- However, `getValueFromVariable`, which `decodeValue` calls internally, [returns the object as is when it receives it](https://github.com/WordPress/gutenberg/blob/33bf984897683adfc4582a29adb079cf6ee0ab59/packages/block-editor/src/components/global-styles/utils.js#L322)
- Therefore, I made sure to decode each property value individually

However, this seems to have resulted in the following two regressions:

- [#51619](https://github.com/WordPress/gutenberg/issues/51619): unintentional switch to custom input when moving the slider.
- [#51891](https://github.com/WordPress/gutenberg/issues/51891): Actual values are output as values, not CSS variables (`var(--wp--preset--spacing--{slug})`)

I have done a lot of research and I suspect that #50956 caused the unintended decoding of the value.

## How?

After some trial and error, I found that if the decoding process is not performed when the value is in user preset format (`var: prefixed`), it works as expected. Or perhaps we should keep the decoding process and add a new encoding process. For example, in `BorderPanel`, [the color values are encoded](https://github.com/WordPress/gutenberg/blob/33bf984897683adfc4582a29adb079cf6ee0ab59/packages/block-editor/src/components/global-styles/border-panel.js#L89C8-L89C25).

## Testing Instructions

Enable EmptyTheme and update `theme.json` as follows:

<details><summary>theme.json</summary>

```
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"settings": {
		"appearanceTools": true,
		"layout": {
			"contentSize": "650px",
			"wideSize": "1200px"
		},
		"spacing": {
			"spacingSizes": [
				{
					"name": "Static Value",
					"size": "8px",
					"slug": "1"
				},
				{
					"name": "Clamp Value",
					"size": "clamp(1rem, 0.818rem + 0.91vw, 1.5rem);",
					"slug": "2"
				},
				{
					"name": "CSS Variable (Custom)",
					"size": "var(--wp--custom--font-size--20)",
					"slug": "3"
				},
				{
					"name": "CSS Variable (Font Size)",
					"size": "var(--wp--preset--font-size--24)",
					"slug": "4"
				}
			]
		},
		"typography": {
			"fontSizes": [
				{
					"name": "24px",
					"size": "24px",
					"slug": "24"
				}
			]
		},
		"custom": {
			"font-size": {
				"20": "20px"
			}
		}
  }
}
```
</details> 

This `theme.json` has many variations in the spacingSizes property.

### Test for #51619

- When the slider is moved, it does not switch to custom input.

<details><summary>Screencast</summary>

https://github.com/WordPress/gutenberg/assets/54422211/69f9c586-8548-4b88-b33d-db0cd6e3bedc

</details> 

### Test for #51891

- When moving the slider, the value is always updated as a CSS variable (`var(--wp--preset--spacing--{slug})`).
- When you enter an arbitrary value in a custom input, it will be updated with that value.
- If you enter a value (`8`) that exists in `spacingSizes` in the custom input, it will be reflected in the slider.

<details><summary>Screencast</summary>

https://github.com/WordPress/gutenberg/assets/54422211/ca0ffe39-e1d7-4fcf-bd4d-f6a5e08cb1ce

</details> 

### Test for #50940

- When spacing is updated in global style, the value of the control is not reset and the position of the slider is maintained.

<details><summary>Screencast</summary>

https://github.com/WordPress/gutenberg/assets/54422211/68595430-f52d-4609-8dad-c1314d604e19

</details>

### CODE REVIEW

File: packages/block-editor/src/components/global-styles/dimensions-panel.js [hunk 1]
```diff
@@ -210,11 +210,15 @@ export default function DimensionsPanel( {
 	const decodeValue = ( rawValue ) => {
 		if ( rawValue && typeof rawValue === 'object' ) {
 			return Object.keys( rawValue ).reduce( ( acc, key ) => {
-				acc[ key ] = getValueFromVariable(
-					{ settings },
-					'',
-					rawValue[ key ]
-				);
+				if ( rawValue[ key ].startsWith( 'var:' ) ) {
```
Comment (andrewserong) [inline packages/block-editor/src/components/global-styles/dimensions-panel.js]: Do we need to add either optional chaining or to check if the value is a string before attempting to call `.startsWith`? Over in [`getValueFromVariable`](https://github.com/WordPress/gutenberg/blob/0bf554d746a0b0f1a3a1b1d98a4fda545efa34a5/packages/block-editor/src/components/global-styles/utils.js#L311), the `startsWith` calls only occur if the value is a string.

**Review by andrewserong — COMMENTED** [review]
What an interesting problem, thanks for diving deep into this one @t-hamano! Apologies, the following comments will be fairly long, hope my thoughts are more useful than they are rambling 😄 

This does appear to fix the issue, the following before / after shows the Before state as containing the "real" values rather than the reference to the variable:

| Before | After |
| --- | --- |
| image | image |

Something I'm curious about though, is why we need to skip calling `getValueFromVariable` for each of these sides 🤔

From a quick look, it seems that the spacing input control attempts to match a value back to its spacing preset here, I'm wondering if there's a problem somehow related to that logic? https://github.com/WordPress/gutenberg/blob/4bdc58146245014f58526922fb1b024a3c49ce50/packages/block-editor/src/components/spacing-sizes-control/input-controls/spacing-input-control.js#L56-L57

In terms of the "real" values accidentally being serialized, I noticed that I can only ever see this happening when some form of axial sides are used. When we go to update a non-axial value (or more specifically, a grouped value like `blockGap` without axial sides), everything seems okay for me without this PR (as far as I can tell), and when I make a change as in the markup of the screenshots above, I see that the side being updated appears to use the `var:` prefix correctly, but it's the _other_ side that winds up using the decoded value. E.g. I update `horizontal` and it gets the correct syntax, but `vertical` gets the wrong one.

So, could it possibly be to do with the logic in `SpacingSizesControl` when we update a particular side, but the remaining sides are passed in without conversion back to the preset syntax? For example, the value we're changing should wind up using the correct syntax [here](https://github.com/WordPress/gutenberg/blob/4bdc58146245014f58526922fb1b024a3c49ce50/packages/block-editor/src/components/spacing-sizes-control/input-controls/spacing-input-control.js#L57), however in the axial logic, the existing (decoded) values are spread and stored for the remaining sides:

https://github.com/WordPress/gutenberg/blob/4bdc58146245014f58526922fb1b024a3c49ce50/packages/block-editor/src/components/spacing-sizes-control/input-controls/axial.js#L23

The same thing appears to happen for individual sides here:

https://github.com/WordPress/gutenberg/blob/4bdc58146245014f58526922fb1b024a3c49ce50/packages/block-editor/src/components/spacing-sizes-control/input-controls/single.js#L19

I guess this gets back to your initial question of whether or not we should skip decoding (as in this PR), or if we should intentionally re-encode the remaining sides that are not the one currently being updated. It could be good to add re-encoding for the remaining sides, to match the implicit re-encoding that's happening for the sides that are being updated. Not sure how complex that'd be to do 🤔

I don't feel too strongly about which way to go, and given the time constraints for 6.3, and that this is quite a small fix code-wise, I think I'd lean slightly toward going with this PR in the short-term as it appears to be testing well. The only change I'd suggest before landing is to add optional chaining for the `.startsWith` call.

What do you think? I know you've spent much longer looking at this than I have, so apologies if I've missed some nuances here 🙂

**Comment by t-hamano** [comment]
Thank you for the detailed survey, @andrewserong!

> So, could it possibly be to do with the logic in `SpacingSizesControl` when we update a particular side, but the remaining sides are passed in without conversion back to the preset syntax?

Yes, perhaps we should explore this a bit more deeply to find the root cause. I will look into this a bit more in the next few days, although we'are under time constraints for the 6.3 release 👀

**Comment by t-hamano** [comment]
> So, could it possibly be to do with the logic in `SpacingSizesControl` when we update a particular side, but the remaining sides are passed in without conversion back to the preset syntax? For example, the value we're changing should wind up using the correct syntax [here](https://github.com/WordPress/gutenberg/blob/4bdc58146245014f58526922fb1b024a3c49ce50/packages/block-editor/src/components/spacing-sizes-control/input-controls/spacing-input-control.js#L57), however in the axial logic, the existing (decoded) values are spread and stored for the remaining sides:

The following changes, which encode the existing (decoded) values, appear to work as expected.

<details><summary>Diff</summary>

```diff
diff --git a/packages/block-editor/src/components/spacing-sizes-control/input-controls/axial.js b/packages/block-editor/src/components/spacing-sizes-control/input-controls/axial.js
index 3ca76d9c8f..6811349b74 100644
--- a/packages/block-editor/src/components/spacing-sizes-control/input-controls/axial.js
+++ b/packages/block-editor/src/components/spacing-sizes-control/input-controls/axial.js
@@ -2,7 +2,12 @@
  * Internal dependencies
  */
 import SpacingInputControl from './spacing-input-control';
-import { LABELS, ICONS, hasAxisSupport } from '../utils';
+import {
+       LABELS,
+       ICONS,
+       getPresetValueFromCustomValue,
+       hasAxisSupport,
+} from '../utils';
 
 const groupedSides = [ 'vertical', 'horizontal' ];
:...skipping...
diff --git a/packages/block-editor/src/components/spacing-sizes-control/input-controls/axial.js b/packages/block-editor/src/components/spacing-sizes-control/input-controls/axial.js
index 3ca76d9c8f..6811349b74 100644
--- a/packages/block-editor/src/components/spacing-sizes-control/input-controls/axial.js
+++ b/packages/block-editor/src/components/spacing-sizes-control/input-controls/axial.js
@@ -2,7 +2,12 @@
  * Internal dependencies
  */
 import SpacingInputControl from './spacing-input-control';
-import { LABELS, ICONS, hasAxisSupport } from '../utils';
+import {
+       LABELS,
+       ICONS,
+       getPresetValueFromCustomValue,
+       hasAxisSupport,
+} from '../utils';
 
 const groupedSides = [ 'vertical', 'horizontal' ];
 
@@ -20,7 +25,17 @@ export default function AxialInputControls( {
                if ( ! onChange ) {
                        return;
                }
-               const nextValues = { ...values };
+
+               // Encode the existing value into the preset value if the passed in value matches the value of one of the spacingSizes.
+               const nextValues = {
+                       ...Object.keys( values ).reduce( ( acc, key ) => {
+                               acc[ key ] = getPresetValueFromCustomValue(
+                                       values[ key ],
+                                       spacingSizes
+                               );
+                               return acc;
+                       }, {} ),
+               };
 
                if ( side === 'vertical' ) {
                        nextValues.top = next;
diff --git a/packages/block-editor/src/components/spacing-sizes-control/input-controls/separated.js b/packages/block-editor/src/components/spacing-sizes-control/input-controls/separated.js
index ddfa3ebca8..e9e2f82880 100644
--- a/packages/block-editor/src/components/spacing-sizes-control/input-controls/separated.js
+++ b/packages/block-editor/src/components/spacing-sizes-control/input-controls/separated.js
@@ -2,7 +2,12 @@
  * Internal dependencies
  */
 import SpacingInputControl from './spacing-input-control';
-import { ALL_SIDES, LABELS, ICONS } from '../utils';
+import {
+       ALL_SIDES,
+       LABELS,
+       ICONS,
+       getPresetValueFromCustomValue,
+} from '../utils';
 
 export default function SeparatedInputControls( {
        minimumCustomValue,
@@ -20,7 +25,17 @@ export default function SeparatedInputControls( {
                : ALL_SIDES;
 
        const createHandleOnChange = ( side ) => ( next ) => {
-               const nextValues = { ...values };
+               // Encode the existing value into the preset value if the passed in value matches the value of one of the spacingSizes.
+               const nextValues = {
+                       ...Object.keys( values ).reduce( ( acc, key ) => {
+                               acc[ key ] = getPresetValueFromCustomValue(
+                                       values[ key ],
+                                       spacingSizes
+                               );
+                               return acc;
+                       }, {} ),
+               };
+
                nextValues[ side ] = next;
 
                onChange( nextValues );
diff --git a/packages/block-editor/src/components/spacing-sizes-control/input-controls/single.js b/packages/block-editor/src/components/spacing-sizes-control/in
put-controls/single.js
index 2bb0a409da..df6beb0f8f 100644
--- a/packages/block-editor/src/components/spacing-sizes-control/input-controls/single.js
+++ b/packages/block-editor/src/components/spacing-sizes-control/input-controls/single.js
@@ -2,7 +2,7 @@
  * Internal dependencies
  */
 import SpacingInputControl from './spacing-input-control';
-import { LABELS } from '../utils';
+import { LABELS, getPresetValueFromCustomValue } from '../utils';
 
 export default function SingleInputControl( {
        minimumCustomValue,
@@ -16,7 +16,17 @@ export default function SingleInputControl( {
        values,
 } ) {
        const createHandleOnChange = ( currentSide ) => ( next ) => {
-               const nextValues = { ...values };
+               // Encode the existing value into the preset value if the passed in value matches the value of one of the spacingSizes.
+               const nextValues = {
+                       ...Object.keys( values ).reduce( ( acc, key ) => {
+                               acc[ key ] = getPresetValueFromCustomValue(
+                                       values[ key ],
+                                       spacingSizes
+                               );
+                               return acc;
+                       }, {} ),
+               };
+
                nextValues[ currentSide ] = next;
 
                onChange( nextValues );
```

</details> 

The only thing that has not yet been resolved is when a CSS variable that automatically generates the font size is specified as a value in the `size` property of `spacingSizes`, as reported in #51619. When I use the `theme.json` I attached at the beginning of this PR and decode each spacing preset with `getValueFromVariable()`, only the spacing preset that specifies the CSS variable that the font size generates is decoded to **the actual value (24px)**.

```javascript
// "size": "8px"
console.log( getValueFromVariable( { settings }, '', 'var:preset|spacing|1' ) );
// > Expected: "8px"
// > ✅ Actual  : "8px"

// "size": "clamp(1rem, 0.818rem + 0.91vw, 1.5rem);"
console.log( getValueFromVariable( { settings }, '', 'var:preset|spacing|2' ) );
// > Expected: "clamp(1rem, 0.818rem + 0.91vw, 1.5rem);"
// > ✅ Actual  : "clamp(1rem, 0.818rem + 0.91vw, 1.5rem);"

// "size": "var(--wp--custom--font-size--20)"
console.log( getValueFromVariable( { settings }, '', 'var:preset|spacing|3' ) );
// > Expected: "var(--wp--custom--font-size--20)"
// > ✅ Actual  : "var(--wp--custom--font-size--20)"

// "size": "var(--wp--preset--font-size--24)"
console.log( getValueFromVariable( { settings }, '', 'var:preset|spacing|4' ) );
// > Expected: "var(--wp--preset--font-size--24)"
// > ❌ Actual  : "24px" No actual value expected
```

As a result, when the slider is moved to that value, the value is saved as a preset value (`var(--wp--preset--spacing--{slug}`)) but switches to a custom input.

https://github.com/WordPress/gutenberg/assets/54422211/e0ada54b-da20-43c6-932d-51ea954c615a

If only this part could be solved, everything might work fine.

**Comment by t-hamano** [comment]
I tried a different approach.

1. In `SpacingSizesControl` component, when any value (uni-directional, horizontal or vertical) is changed, encode the existing value using `getPresetValueFromCustomValue` to match the format of the new changed value.
2. Narrow down the `features` passed to `getValueFromVariable()'s `first argument in `decodeValue` to only those related to spacing (`spacing`/`dimensions`) so that if a preset value in `spacingSizes` refers to another preset value (such as `var(--wp--preset--font-size--{slug})`), it will not resolve to the actual value

As far as I have tested, everything works correctly and I believe this is the logic that should be expected. However, I am a little concerned that the second point, the approach of narrowing down the functionality, may affect other functions 🤔.

**Review by andrewserong — COMMENTED** [review]
Nice progress on this @t-hamano! I think this is honing in on a good direction, and I agree with your assessment that we should only be decoding dimensions and spacing sizes for the dimensions panel, so I think that approach looks good to me 👍

While testing, I noticed an issue with the added calls to `getPresetValueFromCustomValue` which is that `undefined` value unintentionally match against the `0` preset, so if you go to change the top/bottom padding for example, the left/right padding will unexpectedly be set to the `0` preset. This issue isn't _caused_ by this PR per se, but it's exposed by it. I think I've found a fix for it, so I've opened up a separate PR over in https://github.com/WordPress/gutenberg/pull/52711 if you have time to take a look.

I _think_ if we merge that PR and this one, we should be in a good place for the functionality. What do you think?

Thanks again for all the detailed investigation here!

**Comment by t-hamano** [comment]
Thanks for bringing the issue to my attention, @andrewserong!

I would like to rebase and re-test this PR after #52711 is merged. I, too, believe that these two PRs are sufficient solutions 😊

**Review by andrewserong — APPROVED** [review]
This is testing nicely for me now after the rebase, the linked issues are resolved and I haven't encountered any issues with any of the dimensions controls in the post or site editors and global styles. LGTM, thanks again for the work here! ✨

**Comment by t-hamano** [comment]
Thanks for the testing! I have also tried the testing instruction at the beginning of this PR in various ways, and I believe they all work as expected 🚀

**Comment by ramonjd** [comment]
I just cherry-picked this PR to the [update/second-round-RC1](https://github.com/WordPress/gutenberg/pull/52677) branch to get it included in the next release: 77f25c0ab7d6f7091092795fdb45b13a4561b1ef

**Pull request merged**